mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp

```diff
@@ -60,9 +60,51 @@ struct ConcatOptimization : public OpRewritePattern<tosa::ConcatOp> {
   }
 };
 
+struct SelfConcatToTile : public OpRewritePattern<tosa::ConcatOp> {
+  using OpRewritePattern<tosa::ConcatOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::ConcatOp op,
+                                PatternRewriter &rewriter) const override {
+    if (llvm::all_equal(op->getUsers())) {
+      const auto concatUser =
+          llvm::dyn_cast<tosa::ConcatOp>(op->getUses().begin()->getOwner());
+      if (concatUser) {
+        // Try folding the concat into its consumer before rewriting it to a
+        // tile.
+        SmallVector<Value> replacementValues;
+        auto foldResult = rewriter.tryFold(concatUser, replacementValues);
+        if (foldResult.succeeded()) {
+          if (!replacementValues.empty()) {
+            rewriter.replaceOp(concatUser, replacementValues);
+          }
+          return failure();
```

Comment:
Shouldn't this be a success?

Reply:
Good catch. I returned failure as the Pattern did not do the canonicalization and instead we folded, but the IR got still modified